client/consensus/manual-seal/src/rpc.rs

```diff
@@ -0,0 +1,220 @@
+	#[rpc(name = "engine_finalizeBlock")]
+	fn finalize_block(
+		&self,
+		hash: Hash,
+		justification: Option<Justification>
+	) -> FutureResult<bool>;
+}
+
+/// A struct that implements the [`ManualSealApi`].
+pub struct ManualSeal<Hash> {
+	import_block_channel: mpsc::Sender<EngineCommand<Hash>>,
+}
+
+/// return type of `engine_createBlock`
+#[derive(Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub struct CreatedBlock<Hash> {
+	/// hash of the created block.
+	pub hash: Hash,
+	/// some extra details about the import operation
+	pub aux: ImportedAux
+}
+
+impl<Hash> ManualSeal<Hash> {
+	/// Create new `ManualSeal` with the given reference to the client.
+	pub fn new(import_block_channel: mpsc::Sender<EngineCommand<Hash>>) -> Self {
+		Self { import_block_channel }
+	}
+}
+
+impl<Hash: Send + 'static> ManualSealApi<Hash> for ManualSeal<Hash> {
+	fn create_block(
+		&self,
+		create_empty: bool,
+		finalize: bool,
+		parent_hash: Option<Hash>
+	) -> FutureResult<CreatedBlock<Hash>> {
+		let mut sink = self.import_block_channel.clone();
+		let future = async move {
+			let (sender, receiver) = oneshot::channel();
+			let command = EngineCommand::SealNewBlock {
+				create_empty,
+				finalize,
+				parent_hash,
+				sender: Some(sender),
+			};
+			sink.send(command).await.map_err(map_error)?;
+
+			match receiver.await {
+				// all good
+				Ok(Ok(block)) => Ok(block),
+				// error from the authorship task
+				Ok(Err(e)) => {
+					Err(Error {
+						code: ErrorCode::ServerError(500),
+						message: format!("{}", e),
+						data: None,
+					})
+				}
+				// channel has been dropped
+				Err(_) => {
+					Err(Error {
+						code: ErrorCode::ServerError(SERVER_SHUTTING_DOWN),
+						message: "Consensus process is terminating".into(),
+						data: None,
+					})
+				}
+			}
+		}.boxed();
+
+		Box::new(future.compat())
+	}
+
+	fn finalize_block(&self, hash: Hash, justification: Option<Justification>) -> FutureResult<bool> {
+		let mut sink = self.import_block_channel.clone();
+		let future = async move {
+			let (sender, receiver) = oneshot::channel();
+			sink.send(
+				EngineCommand::FinalizeBlock { hash, sender: Some(sender), justification }
+			).await.map_err(map_error)?;
+
+			match receiver.await {
+				// all good
+				Ok(Ok(())) => Ok(true),
+				// error from the authorship task
+				Ok(Err(e)) => {
+					Err(Error {
+						code: ErrorCode::ServerError(500),
+						message: format!("{}", e),
+						data: None
+					})
+				}
+				// channel has been dropped
+				Err(_) => {
+					Err(Error {
+						code: ErrorCode::ServerError(SERVER_SHUTTING_DOWN),
+						message: "Consensus process is terminating".into(),
+						data: None
+					})
+				}
+			}
+		}.boxed();
+
+		Box::new(future.compat())
+	}
+}
+
+/// convert a futures::mpsc::SendError to jsonrpc::Error
+fn map_error(error: SendError) -> Error {
+	if error.is_disconnected() {
+		log::warn!("Received sealing request but Manual Sealing task has been dropped");
+	}
+
+	Error {
```

Comment:
This is being repeated multiple times. Can you instead introduce an `enum` with different errors and just implement `From<Error> for jsonrpc_core::Error` to do the conversion in a single place instead? Like:
```rust
enum Error {
  ConsensusTerminating,
  ...
}
impl From<Error> for jsonrpc_core::Error {
 fn from(e: Error) -> Self {
   match e {
     Error::ConsensusTerminating => jsonrpc_core::Error {
        code: ...,
        ...
     } 
   }
}
```